Fix building sak with cross-compilers

MisterDA opened the pull request:
The build and target compiler invocations may differ. For instance, if cross-compiling with build/host `x86_64-*-linux` and target `x86_64-pc-windows`, the build/host object extension is `.o`, but the target uses `.obj`.

The `AX_PROG_CC_FOR_BUILD` macro has fixes from revision 21 to 26 that are needed, or it would overwrite `OBJEXT` and `EXEEXT` with `BUILD_OBJEXT` and `BUILD_EXEEXT` without defining the latter.

A subtlety is that building on Windows for native Windows targets isn't cross-compiling, thus it's not the build compiler for sak (`*-pc-cygwin`/`*-pc-msys`) that is used, but the host/target compiler.

cc @shym

MisterDA committed:
Prefer Autoconf EXEEXT to our exeext

Autoconf detects `.exe` on all Windows targets. OBJEXT and EXEEXT are
automatically substituted.

MisterDA committed:
Fix sak build not using the proper build compiler invocation

sak doesn't need separate compilation and linking stages, and this
avoids adding build-specific variables to the Makefiles.

Co-authored-by: shym <samuel@tarides.com>

MisterDA committed:
Reformat an AS_IF

gasche commented:
As a non-expert I looked at whether I could close my eyes and merge it right away. My assessment is "not quite":
- the first two commits look fine and I would merge them
- the third commit looks globally okay, and/but it is the first (unless I missed something) to introduce BUILD_* variables at the Makefile level. If I understand correctly, #13526 left Makefiles mostly unchanged (except Makefile.cross which is for advanced stuff), and now we are gradually introducing host/build/target distinctions into normal makefiles. I'm not opposed, but I wouldn't mind getting confirmation from Someone Knowledgeable (@dra27, @shindere, @shym) that this is the right way to go.

shym reviewed (commented):
- b5b1daa94e Prefer Autoconf EXEEXT to our exeext
  LGTM
- b8cdedb102 Reformat an AS_IF
  LGTM
- 034f8b7f80 Fix sak build not using the proper build compiler invocation
  I don’t understand what use cases this is fixing (I see the changes of extensions in cross-compiler setups but is there anything more than aesthetic considerations?).
  I agree that this is introducing quite some variables. To avoid the `$(BUILD_O)` for instance, I would rather go that way:
  - merge the rules for `sak` into one (separating compilation from linking is not really useful there),
  - define a do-it-all `SAK_BUILD` variable in `Makefile.build_config.in` set up during `configure`,
  - if the extension of the `sak` binary really matters (note that we often use `.exe` extensions on Unix...), define `SAK` also in `configure`/`Makefile.build_config.in`.
  
  But I like the idea of defining `outputobj` and `outputexe` at the same place in `configure.ac`!

MisterDA committed:
Update AX_PROG_CC_FOR_BUILD to rev 26

The AX_PROG_CC_FOR_BUILD macro has fixes from revision 21 to 26 that
are needed, or it would overwrite OBJEXT and EXEEXT with BUILD_OBJEXT
and BUILD_EXEEXT without defining the latter.

MisterDA committed:
configure: define outputobj and outputexe flags together

MisterDA commented:
> I don’t understand what use cases this is fixing (I see the changes of extensions in cross-compiler setups but is there anything more than aesthetic considerations?).

``` console
$ clang-20  -Foruntime/sak.o -c runtime/sak.c
$ clang-20  -o runtime/sak runtime/sak.o
clang-20: error: no such file or directory: 'runtime/sak.o'
clang-20: error: no input files
```

I'm using here clang-20 for the build compiler, but with a `*-pc-windows` target (MSVC ccomp_type). You see that the build flags are mixed up: it's using `-Fo`.
The `AX_PROG_CC_FOR_BUILD` update fixes a confusion between the build and the target object and executable extensions.

> I agree that this is introducing quite some variables. To avoid the `$(BUILD_O)` for instance, I would rather go that way:
> 
> * merge the rules for `sak` into one (separating compilation from linking is not really useful there),
> * define a do-it-all `SAK_BUILD` variable in `Makefile.build_config.in` set up during `configure`,
> * if the extension of the `sak` binary really matters (note that we often use `.exe` extensions on Unix...), define `SAK` also in `configure`/`Makefile.build_config.in`.

I agree this could be cleaner, but if we build sak using `MKEXE_VIA_CC`, we can't pass it the list of cflags, because of the msvc syntax (cflags need to be passed before the `/link` flag). This would require some special casing or duplication… I'm scratching my head and I can't find a nice way.

I think dra27 has plans to remove sak, so introducing these build variables could be seen as temporary.

MisterDA commented:
I've managed to make a single invocation to build sak. MSVC still generates a `sak.obj`. That's a bit unfortunate. I have two proposals: one using the direct invocation: [232aa53](https://github.com/ocaml/ocaml/pull/13804/commits/232aa535baf307a60c03bdb4993513778aa4cc2f), and one with build variables [9e09521](https://github.com/MisterDA/ocaml/pull/126/commits/9e095211c19d2f07c401df00dd3dbe99247618e9) (it uses `BUILD_OBJEXT`/`BUILD_O`, `BUILD_EXEEXT`, `build_outputexe` and `build_outputobj`). My preference goes to using build variables.

The `run-crosscompiler-tests` label should be added to this PR

shym commented on configure.ac:
```diff
@@ -1200,6 +1170,32 @@ AS_CASE([$ocaml_cc_vendor,$target],
+
+AS_IF([test x"$build" = x"$target" -o x"$target_runnable" = xtrue],
+  [AS_CASE([$target],
+  [*-pc-windows],
+    [SAK_BUILD=m4_normalize(
+      "\$(CC) \$(OC_CFLAGS) \$(CFLAGS) \$(OC_CPPFLAGS) \$(CPPFLAGS) \$(2)
+      /link /out:\$(1) $oc_exe_ldflags \$(OC_LDFLAGS) \$(LDFLAGS)")],
+  [SAK_BUILD=m4_normalize(
+    '$(MKEXE_VIA_CC) $(OC_CFLAGS) $(CFLAGS) $(OC_CPPFLAGS) $(CPPFLAGS)')])
```
```suggestion
    [*-pc-windows],
      [SAK_BUILD=m4_normalize([
        "\$(CC) \$(OC_CFLAGS) \$(CFLAGS) \$(OC_CPPFLAGS) \$(CPPFLAGS) \$(2)
        /link /out:\$(1) $oc_exe_ldflags \$(OC_LDFLAGS) \$(LDFLAGS)"])],
    [SAK_BUILD=m4_normalize([
      '$(MKEXE_VIA_CC) $(OC_CFLAGS) $(CFLAGS) $(OC_CPPFLAGS) $(CPPFLAGS)'])])
```
Explicit m4 quotes and fixed indentation.

shym commented on configure.ac:
```diff
@@ -1200,6 +1170,32 @@
+  [AC_MSG_NOTICE([detecting the C toolchain for build])
+  AX_PROG_CC_FOR_BUILD
+  # Note that *-pc-windows is not supported for _build_ so we can use '-o'
+  SAK_BUILD=m4_normalize(
+    "$CC_FOR_BUILD $LDFLAGS_FOR_BUILD -o \$(1) \$(2)\
+    $CFLAGS_FOR_BUILD $CPPFLAGS_FOR_BUILD")
```
```suggestion
  SAK_BUILD=m4_normalize(["$CC_FOR_BUILD $LDFLAGS_FOR_BUILD -o \$(1) \$(2)
    $CFLAGS_FOR_BUILD $CPPFLAGS_FOR_BUILD"])
```

shym reviewed (approved):
Thank you for that update!
I prefer that version and it looks good to me, I have only minor stylistic suggestions.

gasche reviewed (approved):
Thanks both!
